clippy_lints/src/manual_checked_sub.rs

```diff
@@ -0,0 +1,239 @@
+            } else if (is_referencing_same_variable(sub_lhs, self.condition_lhs)
+                || are_literals_equal(sub_lhs, self.condition_lhs))
+                && (is_referencing_same_variable(sub_rhs, self.condition_rhs)
+                    || are_literals_equal(sub_rhs, self.condition_rhs))
+            {
+                self.build_suggession(expr);
+            }
+        }
+
+        walk_expr(self, expr);
+    }
+}
+
+impl<'tcx> SubExprVisitor<'_, 'tcx> {
+    fn build_suggession(&mut self, expr: &Expr<'tcx>) {
+        let body_snippet = snippet_with_applicability(self.cx, self.if_body_block.span, "..", &mut self.applicability);
+
+        let binary_expr_snippet = snippet_with_applicability(self.cx, expr.span, "..", &mut self.applicability);
+        let updated_usage_context_snippet = body_snippet.as_ref().replace(binary_expr_snippet.as_ref(), "result");
+        // let else_snippet = snippet_with_applicability(self.cx, self.if_else_block.span, "..", &mut
+        // self.applicability);
+
+        let lhs_snippet = snippet_with_applicability(self.cx, self.condition_lhs.span, "..", &mut self.applicability);
+        let rhs_snippet = snippet_with_applicability(self.cx, self.condition_rhs.span, "..", &mut self.applicability);
+
+        let lhs_needs_parens = matches!(self.condition_lhs.kind, ExprKind::Cast(..));
+
+        let mut suggestion = if lhs_needs_parens {
+            format!("if let Some(result) = ({lhs_snippet}).checked_sub({rhs_snippet}) {updated_usage_context_snippet}")
+        } else {
+            format!("if let Some(result) = {lhs_snippet}.checked_sub({rhs_snippet})  {updated_usage_context_snippet}")
+        };
+
+        if let Some(else_expr) = self.else_block {
+            let else_snippet = snippet_with_applicability(self.cx, else_expr.span, "..", &mut self.applicability);
+            // suggestion.push_str(&format!(" else {else_snippet}"));
+            write!(suggestion, " else {else_snippet}").unwrap();
+        }
+
+        span_lint_and_sugg(
+            self.cx,
+            MANUAL_CHECKED_SUB,
+            self.if_expr.span,
+            "manual re-implementation of checked subtraction",
+            "consider using `.checked_sub()`",
+            suggestion,
+            self.applicability,
+        );
+    }
+}
+
+fn is_unsigned_int<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) -> bool {
+    let expr_type = cx.typeck_results().expr_ty(expr).peel_refs();
+    if matches!(expr_type.kind(), ty::Uint(_)) {
+        return true;
+    }
+
+    if let ExprKind::Lit(lit) = &expr.kind {
+        if let LitKind::Int(_, LitIntType::Unsuffixed) = &lit.node {
+            return true;
+        }
+    }
+    false
+}
+
+fn are_literals_equal<'tcx>(expr1: &Expr<'tcx>, expr2: &Expr<'tcx>) -> bool {
+    if let (ExprKind::Lit(lit1), ExprKind::Lit(lit2)) = (&expr1.kind, &expr2.kind) {
+        if let (LitKind::Int(val1, suffix1), LitKind::Int(val2, suffix2)) = (&lit1.node, &lit2.node) {
```

Comment:
Those two `if let` can be collapsed:

```suggestion
    if let (ExprKind::Lit(lit1), ExprKind::Lit(lit2)) = (&expr1.kind, &expr2.kind)
        && let (LitKind::Int(val1, suffix1), LitKind::Int(val2, suffix2)) = (&lit1.node, &lit2.node) {
```